Injector docs update (#105)
silverstripe/developer-docs: en/02_Developer_Guides/05_Extending/05_Injector.md

developers. 

## Basic usage

The following snippet shows `Injector` creating a new object of type `App\MyClassName` through its `create` method:

```php
use App\MyClassName;
use SilverStripe\Core\Injector\Injector;

$object = Injector::inst()->create(MyClassName::class);
```

Repeated calls to `create()` create a new object each time.

```php
use App\MyClassName;
use SilverStripe\Core\Injector\Injector;

$object = Injector::inst()->create(MyClassName::class);
$object2 = Injector::inst()->create(MyClassName::class);

echo $object !== $object2;

// returns true;
```

### Singleton Pattern

The `Injector` API can be used for the [singleton pattern](https://en.wikipedia.org/wiki/Singleton_pattern) through `get()`. Unlike `create()` subsequent calls to `get` return the same object instance as the first call.

```php
use App\MyClassName;
use SilverStripe\Core\Injector\Injector;

// sets up MyClassName as a singleton
$object = Injector::inst()->get(MyClassName::class);
$object2 = Injector::inst()->get(MyClassName::class);

echo ($object === $object2);

// returns true;
```

## Basic dependency injection

> The benefit of constructing objects this way is that the object that injector returns for `ClassName` can be changed by subsequent code or configuration, for example:

Reviewer: ```suggestion
The benefit of constructing objects this way is that the object that the injector returns for `My ClassName` can be changed by subsequent code or configuration, for example:
```

```php
use App\MyClient;
use SilverStripe\Core\Injector\Injector;

// default client created - could be in core code
Injector::inst()->registerService(new ReadClient(), MyClient::class);
$client = Injector::inst()->get(MyClient::class);
// $client is an instance of ReadClient

// somewhere later, perhaps in some application code
Injector::inst()->registerService(new WriteClient(), MyClient::class);
$client = Injector::inst()->get(MyClient::class);
// $client is now an instance of WriteClient

```

[info]Note that 'MyClient' [does not have to be an existing class](#service-inheritance) - you could use an abitrary string to identify it. That said using existing classes can be easier to reason about and can be refactored by automatic tools/IDEs.[/info]

Using Injector imperatively like this is most common [in testing](#testing-with-injector). 

# Injector API 🤝 Configuration API

> The Injector API combined with the Configuration API is a powerful way to declare and manage dependencies in your code. For example  `ClassName` can be swapped out using the following config:

Reviewer: ```suggestion
The Injector API combined with the Configuration API is a powerful way to declare and manage dependencies in your code. For example, `MyClassName` can be swapped out using the following config:
```

**app/_config/app.yml**

```yml
SilverStripe\Core\Injector\Injector:
  App\MyClassName:
    class: MyBetterClassName
```
then used in php:
```php
use App\MyClassName;
use SilverStripe\Core\Injector\Injector;

// sets up MyClassName as a singleton
$object = Injector::inst()->get(MyClassName::class);
// $object is an instance of MyBetterClassName
```

> This allows you to concisely override objects in Silverstripe core or other third party Silverstripe code.

Reviewer: ```suggestion
This allows you to concisely override classes in Silverstripe core or other third-party Silverstripe code.
```

[Wikipedia](https://en.wikipedia.org/wiki/Dependency_injection)

In Silverstripe a combination of the [Injector API](#Injector) and the [Configuration API](../configuration) provide a comprehensive dependency injection pattern. 
Some of the goals of dependency injection are:

* Simplified instantiation of objects
* Providing a uniform way of declaring and managing inter-object dependencies
* Promoting abstraction of logic

In practical terms it allows developers to: 

* Make class dependencies configurable rather than hard-coded
* Override or replace core behaviour without needing to alter core code
* Write more testable code

# Injector

The [Injector](api:SilverStripe\Core\Injector\Injector) class is the central manager of inter-class dependencies in Silverstripe CMS. It offers developers the 
ability to declare the dependencies a class type has, or to change the nature of the dependencies defined by other 
developers. 

## Basic usage

The following snippet shows `Injector` creating a new object of type `MyClassName` through its `create` method:

```php

> use SilverStripe\Core\Injector\Injector;
> 
> $object = Injector::inst()->create('MyClassName');

Reviewer: How about using the `::class` notation throughout in the code examples?

```suggestion
use App\MyClassName;
use SilverStripe\Core\Injector\Injector;

$object = Injector::inst()->create(MyClassName::class);
```

Author: yeah good point. I guess it doesn't have to be a class but using `::class` does feel more like a 'real' use case. I could use that and make sure there is a reference to being able to use non-class strings.

Reviewer: In the yaml config examples it would have to be the string version of the fully namespaced class name, so we can keep it there. I think it's also good to use a namespaced class in the yaml example, too, as people often forget to use the namespaces there.

Author: yeah all good - should now be using that syntax 👍

      FilesystemAdapter: '%$SilverStripe\Assets\Flysystem\ProtectedAdapter'
```

#### Using constants and environment variables

The Injector configuration has the special ability to include core constants or environment variables. They can be used by quoting with back ticks "`". Please ensure you also quote the entire value (see below).

```yaml
SilverStripe\Core\Injector\Injector:
  CachingService:
    class: SilverStripe\Cache\CacheProvider
    properties:
      CacheDir: '`TEMP_DIR`'
```

Environment variables are used in the same way:

```yml
SilverStripe\Core\Injector\Injector:
  App\Services\MyService:
      class: App\Services\MyService
      constructor:
        baseURL: '`SS_API_URI`'
        credentials:
          id: '`SS_API_CLIENT_ID`'

>           secret: '`SS_API_CLIENT_SECRET`'

Reviewer: Inconsistent indentation

Author: not 100% sure what you mean here sorry?

Reviewer: seems like from the class down it's 2 spaces more than it should be

```

Note: undefined variables will be replaced with null.
## Dependencies

Silverstripe classes can delcare a special `$dependencies` array which can quickly configure dependencies when used with Injector. Injector will create instantiate an object for every array value and assign it to a property that matches the array key. For example: 

```php
namespace App;
use SilverStripe\Control\Controller;
use ThirdParty\PermissionService;

class MyController extends Controller 
{

    // properties matching the array keys in $dependencies will be automatically
    // set by the injector on object creation. Note it must be a public property.
    public $textProperty;

    // we declare the types for each of the properties on the object. Anything we pass in via the Injector API must
    // match these data types.
    private static $dependencies = [

>         'permissions'        => '%$'. PermissionService::class,

Reviewer: ```suggestion
        'permissions' => '%$'. PermissionService::class,
```

      FilesystemAdapter: '%$SilverStripe\Assets\Flysystem\ProtectedAdapter'
```

#### Using constants and environment variables

The Injector configuration has the special ability to include core constants or environment variables. They can be used by quoting with back ticks "`". Please ensure you also quote the entire value (see below).

```yaml
SilverStripe\Core\Injector\Injector:
  CachingService:
    class: SilverStripe\Cache\CacheProvider
    properties:
      CacheDir: '`TEMP_DIR`'
```

Environment variables are used in the same way:

```yml
SilverStripe\Core\Injector\Injector:
  App\Services\MyService:
    class: App\Services\MyService
    constructor:
      baseURL: '`SS_API_URI`'
    credentials:
      id: '`SS_API_CLIENT_ID`'
      secret: '`SS_API_CLIENT_SECRET`'
```

> [info]Note: undefined variables will be replaced with null.[/info]
> ## Dependencies

Reviewer: ```suggestion
[info]Note: undefined variables will be replaced with null.[/info]

## Dependencies
```

> Silverstripe classes can delcare a special `$dependencies` array which can quickly configure dependencies when used with Injector. Injector will create instantiate an object for every array value and assign it to a property that matches the array key. For example: 

Reviewer: ```suggestion
Silverstripe classes can declare a special `$dependencies` array which can quickly configure dependencies when used with Injector. Injector will instantiate an object for every array value and assign it to a property that matches the array key. For example: 
```

    // match these data types.
    private static $dependencies = [
        'permissions' => '%$'. PermissionService::class,
    ];
}
```
[info]Note the properties set by Injector must be public properties.[/info]

When creating a new instance of `App\MyController` via Injector the permissions property will contain an instance of the `ThirdParty\PermissionService` that was resolved by Injector.

```php
use SilverStripe\Core\Injector\Injector;

$object = Injector::inst()->get(App\MyController::class);

echo ($object->permissions instanceof ThirdParty\PermissionService);
// returns true;
```

We can then change or override any of those dependencies via the [Configuration YAML](../configuration) and Injector does the hard work of wiring it up.

**app/_config/app.yml**

```yml
SilverStripe\Core\Injector\Injector:
  ThirdParty\PermissionService:
    class: App\MyCustomPermissionService
```

Now the dependencies will be replaced with our configuration.

```php
use SilverStripe\Core\Injector\Injector;

$object = Injector::inst()->get(App\MyController::class);

echo ($object->permissions instanceof App\MyCustomPermissionService);
// returns true;
```

#### Properties

> Injector's configuration can also be used to define properties for example:

Reviewer: ```suggestion
Injector's configuration can also be used to define properties, for example:
```

```yaml
SilverStripe\Core\Injector\Injector:
  App\MyController:
    properties:
      textProperty: 'My Text Value'
```

```php
$object = Injector::inst()->get(App\MyController::class);

echo (is_string($object->textProperty));
// returns true;
```
### Dependent Calls
As well as properties, method calls the class depends on can also be specified via the `calls` property in yaml:

```yml
SilverStripe\Core\Injector\Injector:
  App\Logger:
    class: Monolog\Logger
    calls:

>       - [ pushHandler, [ %$App\Log\DefaultHandler ] ]

Reviewer: ```suggestion
      - [pushHandler, ['%$App\Log\DefaultHandler']]
```

Assuming a class structure such as

```php
namespace App;

class MyController 
{

>   public $permssions;

Reviewer: ```suggestion
  public $permissions;
```

```yml
---

> name: App\MyController

Reviewer: not sure how backslashes work in yaml section names.. would remove it here or put the name in quotes
